src/ArchaeologistApi.ts

```diff
@@ -0,0 +1,333 @@
+   *
+   * @param arch - The archaeologist to hang up.
+   */
+  async hangUp(arch: ArchaeologistData) {
+    // @ts-ignore
+    return this.p2pNode?.hangUp(this.getDialAddress(arch));
+  }
+
+  /**
+   * Pings the given archaeologist. Completes with the latency of the ping. If the archaeologist
+   * cannot be pinged, completes with null.
+   *
+   * @param arch - The archaeologist to ping.
+   * @param onComplete - Callback function to be called after the archaeologist has been pinged. Latency is passed as an argument.
+   * @param pingTimeout - The timeout for the ping.
+   *
+   * @throws if the archaeologist cannot be dialed.
+   * @throws if the archaeologist does not have a valid peerId ("<domain>:<p2pPeerId>").
+   */
+  async pingArchaeologist(
+    arch: ArchaeologistData,
+    onComplete: (latency: number | null) => void,
+    pingTimeout: number = 5000
+  ) {
+    const couldNotConnect = setTimeout(() => {
+      console.log(`${arch.toString()}: ping timeout!`);
+      onComplete(null);
+    }, pingTimeout);
+
+    const peerIdString = arch.profile.peerId;
+    console.log(`pinging ${peerIdString}`);
+
+    // @ts-ignore
+    const latency = await this.p2pNode?.ping(this.getDialAddress(arch));
+    await this.hangUp(arch);
+
+    if (!!latency) {
+      clearTimeout(couldNotConnect);
+      onComplete(latency);
+    }
+  }
+
+  // TODO: Might more sense to have this method operate of a single archaeologist instead,
+  //       and have the caller worry about synchoronizing the results and retry logic
+  /**
+   * Initiates a negotiation with the given archaeologists. Completes with a map of the archaeologists'
+   * signatures and public keys, or with a map of the archaeologists' exceptions if the archaeologists
+   * decline the negotiation.
+   * 
+   * @param selectedArchaeologists - The archaeologists to negotiate with.
+   * @param isRetry - Whether or not this is a retry of a previous negotiation.
+   * 
+   * @returns A map of the archaeologists' signatures and public keys, or a map of the archaeologists' exceptions.
+   * 
+   * @throws if the archaeologists do not have valid peerIds ("<domain>:<p2pPeerId>").
+   * @throws if the archaeologists' streams throw an exception.
+   */
+  async initiateSarcophagusNegotiation(selectedArchaeologists: ArchaeologistData[], isRetry = false) {
+    console.log('starting the negotiation');
+    const lowestRewrapInterval = getLowestRewrapInterval(selectedArchaeologists);
+    const lowestResurrectionTime = getLowestResurrectionTime(selectedArchaeologists);
+
+    const negotiationTimestamp = (await getCurrentTimeSec(this.signer.provider!)) * 1000;
+
+    let archaeologistSignaturesCount = 0;
+    let archaeologistPublicKeysCount = 0;
+    const negotiationResult = new Map<string, ArchaeologistNegotiationResult>([]);
+
+    await Promise.all(
+      selectedArchaeologists.map(async arch => {
+        // if (cancelToken.cancelled) return;
```

Comment:
Need to figure out another way to cancel this process after it has already started. TODO above might not work, because negotiation needs to happen for ALL archaeologists on the sarcophagus (lowest rewrap interval and lowest res time). Might be possible to calculate these values prior on client side and pass them into a single arch negotiation function?

Reply:
Resolved: removed this entirely. Consumer will not be able to cancel negotiation once it starts.